pkg/server/authorizedentries/cache_test.go

```diff
@@ -0,0 +1,441 @@
+	// across various operations. The motivation is to ensure that as the cache
+	// is updated that we are appropriately inserting and removing records from
+	// the indexees.
+	t.Run("pristine", func(t *testing.T) {
+		cache := NewCache()
+		require.Zero(t, cache.stats())
+	})
+
+	t.Run("entries and aliases", func(t *testing.T) {
+		entry1 := makeWorkload(agent1)
+		entry2a := makeWorkload(agent2)
+
+		// Version b will change to an alias instead
+		entry2b := makeAlias(alias1, sel1, sel2)
+		entry2b.Id = entry2a.Id
+
+		cache := NewCache()
+		cache.UpdateEntry(entry1)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  1,
+			EntriesByParentID: 1,
+		}, cache.stats())
+
+		cache.UpdateEntry(entry2a)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  2,
+			EntriesByParentID: 2,
+		}, cache.stats())
+
+		cache.UpdateEntry(entry2b)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  1,
+			EntriesByParentID: 1,
+			AliasesByEntryID:  2, // one for each selector
+			AliasesBySelector: 2, // one for each selector
+		}, cache.stats())
+
+		cache.RemoveEntry(entry1.Id)
+		require.Equal(t, cacheStats{
+			AliasesByEntryID:  2, // one for each selector
+			AliasesBySelector: 2, // one for each selector
+		}, cache.stats())
+
+		cache.RemoveEntry(entry2b.Id)
+		require.Zero(t, cache.stats())
+	})
+
+	t.Run("agents", func(t *testing.T) {
+		cache := NewCache()
+		cache.UpdateAgent(agent1.String(), now.Add(time.Hour), []*types.Selector{sel1})
+		require.Equal(t, cacheStats{
+			AgentsByID:        1,
+			AgentsByExpiresAt: 1,
+		}, cache.stats())
+
+		cache.UpdateAgent(agent2.String(), now.Add(time.Hour*2), []*types.Selector{sel2})
+		require.Equal(t, cacheStats{
+			AgentsByID:        2,
+			AgentsByExpiresAt: 2,
+		}, cache.stats())
+
+		cache.UpdateAgent(agent2.String(), now.Add(time.Hour*3), []*types.Selector{sel2})
+		require.Equal(t, cacheStats{
+			AgentsByID:        2,
+			AgentsByExpiresAt: 2,
+		}, cache.stats())
+
+		cache.RemoveAgent(agent1.String())
+		require.Equal(t, cacheStats{
+			AgentsByID:        1,
+			AgentsByExpiresAt: 1,
+		}, cache.stats())
+
+		cache.RemoveAgent(agent2.String())
+		require.Zero(t, cache.stats())
+	})
+}
+
+func testCache() *cacheTest {
+	return &cacheTest{
+		entries: make(map[string]*types.Entry),
+		agents:  make(map[spiffeid.ID]agentInfo),
+	}
+}
+
+type cacheTest struct {
+	entries map[string]*types.Entry
+	agents  map[spiffeid.ID]agentInfo
+}
+
+type agentInfo struct {
+	ExpiresAt time.Time
+	Selectors []*types.Selector
+}
+
+func (a *cacheTest) pickAgent() spiffeid.ID {
+	for agent := range a.agents {
+		return agent
+	}
+	return spiffeid.ID{}
+}
+
+func (a *cacheTest) withEntries(entries ...*types.Entry) *cacheTest {
+	for _, entry := range entries {
+		a.entries[entry.Id] = entry
+	}
+	return a
+}
+
+func (a *cacheTest) withAgent(node spiffeid.ID, selectors ...*types.Selector) *cacheTest {
+	expiresAt := now.Add(time.Hour * time.Duration(1+len(a.agents)))
+	a.agents[node] = agentInfo{
```

Comment:
is not it preventing to test cases where we have 2 agents with the same time? may we test that scenario?

Reply:
Restructured the test to test this.